torch/csrc/jit/testing/file_check.cpp

```diff
@@ -133,32 +156,124 @@ struct FileCheckImpl {
     doChecks(std::make_shared<std::string>(test_file));
   }
 
-  TORCH_API void addCheck(
-      CheckType type,
-      const std::string& s,
-      c10::optional<size_t> count = c10::nullopt) {
-    Check check(type, s, std::move(count));
+  TORCH_API void run(
+      const std::string& checks_file,
+      const std::string& test_file) {
+    auto checks_ptr = std::make_shared<std::string>(checks_file);
+    parseStrings(checks_ptr);
+    run(test_file);
+  }
 
+  TORCH_API void addCheck(Check check) {
     // consecutive CHECK_DAGs & CHECK_NOTs need to be evaluated as a group
-    if (groups.size() == 0 || (type != CHECK_NOT && type != CHECK_DAG)) {
+    if (groups.size() == 0 ||
+        (check.type_ != CHECK_NOT && check.type_ != CHECK_DAG)) {
       groups.push_back({check});
     } else {
       auto& last_group = groups.back();
-      if (last_group.at(0).type_ == type) {
+      if (last_group.at(0).type_ == check.type_) {
         last_group.push_back(check);
       } else {
         groups.push_back({check});
       }
     }
-
     has_run = false;
   }
 
+  TORCH_API void addCheck(
+      CheckType type,
+      const std::string& s,
+      c10::optional<size_t> count = c10::nullopt) {
+    addCheck(Check(type, s, std::move(count)));
+  }
+
   bool has_run = false;
 
   friend std::ostream& operator<<(std::ostream& out, const FileCheckImpl& fc);
 
  private:
+  void parseStrings(std::shared_ptr<std::string>& checks_file) {
+    std::vector<Check> operands;
+    const static std::vector<std::pair<CheckType, std::string>> check_pairs = {
+        {CHECK, ": "},
+        {CHECK_NEXT, "-NEXT: "},
+        {CHECK_SAME, "-SAME: "},
+        {CHECK_NOT, "-NOT: "},
+        {CHECK_DAG, "-DAG: "},
+        {CHECK_COUNT, "-COUNT-"}, // needs special parsing
+    };
+    const std::string prefix = "# CHECK";
+
+    size_t start = 0;
+    size_t num_checks = 0;
+    start = checks_file->find(prefix, start);
+
+    while (start != std::string::npos) {
```

Comment:
This loops seems complicated. Could we refactor it a bit? Extracting out a function for processing one line is one possible change we can make here for instance.